### ISSUE TITLE
Music stops for some seconds with music library on a bumpy network filesystem

### ISSUE THREAD

**Issue by fungs** [owntone/owntone-server#1310]
Hey, I'm running forked-daapd 26.4+dfsg1-1 on Debian with a large music library on a network file system. The corresponding file server sometimes needs a little to serve the requested data, for example, if the hard disks spin down, or if it is busy with other processes, I believe. When I stream the music to AirPlay speakers, I sometimes get breaks of several seconds in the playback.

I wonder if there is a way to configure a larger look-ahead cache for the media server to read the files. The network connection is quite good otherwise, I usually stream HD videos using the same setup.

Of course I can try to tweak the filesystem driver to do opportunistic caching when a file is opened. However, the player would be the more appropriate place to configure the read cache.

**Comment by ejurgensen** [comment]
Do you have the option of building a newer version and trying with that? Unfortunately the Debian version has fallen quite out of date. I believe current OwnTone reads more ahead, so I think there is a fair chance that it would be better.

**Comment by fungs** [comment]
If you are suggesting this, then yes. I'm currently just testing the setup in a container to see, if there are issues. My plan was to rebuild it with newest software once it has proven to work. @ejurgensen do you have any link to verify this (like source diff or so)?

**Comment by ejurgensen** [comment]
Not sure what kind of verification you mean?

**Comment by fungs** [comment]
Like the buffer size constant in source code that has changed.

**Comment by ejurgensen** [comment]
Ok, but the changes aren't as simple as that. Since 26.4 the player has been basically rewritten, and that also means that the way sources are opened and read has changed.

If you have built yourself, and the problem persists, you can experiment by increasing the buffer size further. It [is this parameter](https://github.com/owntone/owntone-server/blob/master/src/input.c#L44).

**Comment by fungs** [comment]
Ok, I will give it a try once a have some spare time and will report it back here after testing. Thanks for your help so far!

**Comment by fungs** [comment]
I tried with owntone 28.2 and and I still get large gaps in the playback. I believe this is caused by the caching and data request behavior of the network filesystem. I larger or full-file cache within owntone could solve the issue. The network filesystem is fuse-seadrive with a storage backend in the local network.

**Comment by fungs** [comment]
Just read your last sentence: I will increase that parameter and rebuild and test.

**Comment by ejurgensen** [comment]
Does the log say anything about "Source is not providing sufficient data, temporarily suspending playback"? If not, then the issue may be on the output side, so not the network drive.

**Comment by fungs** [comment]
Ok, I did some testing and increasing that parameter up to an equivalent of 20 seconds does not solve the problem. I also increased the log level to the max but never see that "Source is not providing sufficient data, temporarily suspending playback" message either. Is there anything else I could do?

**Comment by ejurgensen** [comment]
Some suggestions:
1. If your speakers support it, try using Airplay 2 instead of 1 (see the setting "raop_disable"). This will mean that the audio is losslessly compressed, thus reducing bandwidth requirements significantly.
2. Set the log level to debug and check what the log says when the issue occurs. If there is packet loss then the speakers normally ahead of time start asking for packet retransmission, which is logged. Other possible issues should also be logged, e.g. if the server is too busy.

**Comment by fungs** [comment]
Airplay 2 does not seem to be supported, unfortunately. The only debug messages I see when music stops playing is:

    [2021-09-14 10:32:34] [DEBUG]     raop: send_progress: Sending SET_PARAMETER to 'Multi'

The output speakers are a bundle of Urbanears Baggen/Stammen which provide a virtual multi-room speaker called multi. They otherwise work great.

**Comment by ejurgensen** [comment]
If this message is there consistently then it might be worth testing with the send_progress message disabled. I will get back to you on how to do that.

Does the problem also appear when you play to a single speaker? So without the virtual multi-room target.

**Comment by fungs** [comment]
Will test with a single speaker for a while now. Those gaps are only appearing once or twice each hour...

**Comment by fungs** [comment]
After testing a while, I believe it's a problem with the multi-room receiver. How can I disable the send_progress message for testing?

**Comment by ejurgensen** [comment]
Yes sorry forgot about it. I’m away these days, so can’t test, but I think you can add a line with `return 0;` here: https://github.com/owntone/owntone-server/blob/63c3d59d2f821e8a872b67b609382333ebc51a4a/src/outputs/raop.c#L2309

**Comment by fungs** [comment]
Sure I'll try that! thx

**Comment by fungs** [comment]
Thanks for your help, unfortunately the patch does not solve the problem. I need to investigate a little in order to rule out issues with either the network connection or the receiver implementation. I'll put this on hold in the meanwhile.

**Comment by fungs** [comment]
Thanks for your help, unfortunately the patch does not solve the problem. I need to investigate a little in order to rule out issues with either the network connection or the receiver implementation. I'll put this on hold in the meanwhile.

**Comment by fungs** [comment]
Thanks @ejurgensen  for all your help and suggestions, I was able to hunt the problem down to a faulty network cable. Owntone works quite well now, keep up the great work!

**Comment by ejurgensen** [comment]
Good to hear, thanks for the notice!
